kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/RS3KVFlushManager.java

```diff
@@ -0,0 +1,217 @@
+      final int kafkaPartition,
+      final PssPartitioner pssPartitioner
+  ) {
+    this.storeId = Objects.requireNonNull(storeId);
+    this.rs3Client = Objects.requireNonNull(rs3Client);
+    this.lssId = Objects.requireNonNull(lssId);
+    this.table = Objects.requireNonNull(table);
+    this.writtenOffsets = Objects.requireNonNull(writtenOffsets);
+    this.kafkaPartition = kafkaPartition;
+    this.pssPartitioner = Objects.requireNonNull(pssPartitioner);
+  }
+
+  @Override
+  public String tableName() {
+    return table.name();
+  }
+
+  @Override
+  public TablePartitioner<Bytes, Integer> partitioner() {
+    return new TablePartitioner<Bytes, Integer>() {
+      @Override
+      public Integer tablePartition(int kafkaPartition, Bytes key) {
+        return pssPartitioner.pss(key.get(), kafkaPartition);
+      }
+
+      @Override
+      public Integer metadataTablePartition(int kafkaPartition) {
+        throw new UnsupportedOperationException();
+      }
+    };
+  }
+
+  @Override
+  public RemoteWriter<Bytes, Integer> createWriter(
+      final Integer pssId,
+      final long consumedOffset
+  ) {
+    if (writers.containsKey(pssId)) {
+      throw new IllegalStateException("already created writer for pss " + pssId);
+    }
+    final var maybeWrittenOffset = writtenOffsets.get(pssId);
+    if (maybeWrittenOffset.isPresent()) {
+      final var writtenOffset = maybeWrittenOffset.get();
+      if (writtenOffset >= consumedOffset) {
+        // This can happen after a restart, during restore because we restore from
+        // the earliest written offset over all pss. For most pss we are restoring
+        // from an offset earlier than the pss's last written offset. In that case,
+        // we wait until we have crossed the written offset for the pss before sending
+        // actual records.
+        // TODO: we should be stricter when checking this case
+        return new NoopWriter(kafkaPartition);
+      }
+    }
+    final var writer = new RS3KVWriter(
+        storeId,
+        rs3Client,
+        table,
+        pssId,
+        lssId,
+        consumedOffset,
+        writtenOffsets.get(pssId),
+        kafkaPartition
+    );
+    writers.put(pssId, writer);
+    return writer;
+  }
+
+  @Override
+  public RemoteWriteResult<Integer> postFlush(long consumedOffset) {
+    for (final var entry : writers.entrySet()) {
+      writtenOffsets.put(entry.getKey(), Optional.of(entry.getValue().endOffset()));
+    }
+    writers.clear();
+    return super.postFlush(consumedOffset);
+  }
+
+  Optional<Long> writtenOffset(final int pssId) {
+    return writtenOffsets.get(pssId);
+  }
+
+  @Override
+  public RemoteWriteResult<Integer> updateOffset(long consumedOffset) {
+    // no op for rs3 store. offsets are updated by RS3Client.writeWalSegment
+    return RemoteWriteResult.success(null);
+  }
+
+  @Override
+  public String failedFlushInfo(long batchOffset, Integer failedTablePartition) {
```

Comment:
```suggestion
  public String failedFlushInfo(final long batchOffset, final Integer failedTablePartition) {
```

Reply:
etc throughout the PR. 

We should probably just add a checkstyle rule for this